Lab2 Bubble & Insertion sort

CameronSchafer opened the pull request:
Fixes #188

Review of colleague's PR #191

#### Changes proposed in this PR:

- Add bubble sort, Insertion sort
- Test bubble sort, Insertion sort

Commit:
Create lab2 project files main.go and main_test.go

Create function bubble in main.go
Update function main to call bubble
Update main test in main_test.go to check output of array

Add bubbleLoop function
Call bubbleLoop function in bubble function
Add bubbleLoop test case to main_test.go

Add loop for bubble sorting functionality

Rename main.go to bubble.go
Rename main_test.go to bubble_test.go

Refactor bubble sort into single function
Refactor bubble_test to match new bubble function

Create folder for merge and bubble sort
Move files into corresponding folders
Add main files for merge sort and merge test

Modify test descriptions for bubble sort test cases
Add insertion sort
Add insertion sort unit tests

Fix spelling mistake in comment
Add logic for merge sort
Add unit tests for merge sort

Change package names

Revert package name back to main for each sort type

Test different package names to please GoLangCI

Revert changes for bubble sort

Refactor code into single file.
Refactor unit tests into single file

Lab 1 - Fibonacci and Negafibonacci (#130)

* Implement Fibonacci and Negafibonacci series
* Add unit tests
* Add code changes requested by review
* Refactored tests as table driven tests

Update test cases to use testify

Make bubble sort, sort a copy of the slice
Clean comments

Delete main.go

Delete main_test.go

Lab 1 - Fibonacci and Negafibonacci (#130)

* Implement Fibonacci and Negafibonacci series
* Add unit tests
* Add code changes requested by review
* Refactored tests as table driven tests

Implement function plus table tests

Create base fib.go file
Create base fib_test.go file

Modify test expected output to match that of the correct output
Add functionality for fib and negafib sequences

Optimise the for loop by removing the multiplication from condition

Implement Fibonacci series - Lab01

Provide Fibonacci and Negafibonacci functions.
Provide dispatcher to return proper series base on input.
Provide tests for implementation.

Implement changes from CodeReview

Unify tests structure
Remove dependency on math package
Fix typo in comment

Completed lab04 - numeronyms

Implement numeronyms
Implement unit tests

CR fixes

Store len in wariable istrad of invoking function many times
Shadow variables in

ci: Use go 1.12 on travis but not golangci-lint

Update the travis config to use version 1.12.x of Go.

Update the golangci-lint version to v1.16.0, which is compatible
with go 1.12. Do this is both .travis.yml and .golangci.yml - the
latter is used by the direct GitHub integration with https://golangci.com.
However, the golangci.com integration still uses Go 1.11.X and there is
no way to tell it to use a later version. This means there may be GitHub
check errors for the linter when using Go 1.12 library methods, but that
is not a required check so it won't block merging. This is the best we
can do until the golangci.com service runs with Go 1.12.

The version of Go we say to use in the README.md is Go 1.12, but the CI
and linter uses Go 1.11. Oops. Fix this to use 1.12.

According to https://docs-staging.travis-ci.com/user/languages/go/#specifying-a-go-version-to-use
any tagged version of Go is usable by travis, but doesn't say what
"tagged" means.  It does tell us it uses "gimme" to install go, and
it looks at https://golang.org/dl/ for available versions. So 1.12.x
is a valid version as of now (stable is go1.12.5).

We could drop the version completely and just always get the latest
stable version by default, but we want to ensure that the travis Go
version and the golangci-lint Go version are the same(-ish).
golangci-lint highly recommends a specific version be installed so the
behaviour does not change unexpectedly so builds are repeatable
(https://github.com/golangci/golangci-lint/blob/master/README.md).

Use a string for the travis Go version as certain versions (1.10 for
example) are parsed as numbers which change the version (1.10 becomes
1.1). So just always use strings.

golangci-lint v1.16.0 is the current latest version according to the
github releases page for it
(https://github.com/golangci/golangci-lint/releases/tag/v1.16.0)

Issue: #198

Lab1 - Implement function to print fibonacci series

Includes the following changes
 - Simple application that prints a fibonacci series of length 7
 - Add function to print fibonacci series supporting negative sequences
 - Add unit tests to cover various sequence lengths

Lab1 - Use Replace instead of ReplaceAll

This avoids an issue we are experiencing with golangci.
golangci does not seem be aware of 1.12 functions.

Replace is available in previous version of golang.
It does not cause any issues with golangci

    Lab 1 - Fibonacci

    * Implement Fibonacci
    * Add unit tests
    * Add test coverage report

brvishnu92 commented on 02_bubble/CameronSchafer/bubble_test.go:
```diff
@@ -0,0 +1,119 @@
+		// t.Run creates a sub test and runs it like a normal test
+		t.Run(test.description, func(t *testing.T) {
+			resultArr := bubble(test.input)
+			if !reflect.DeepEqual(resultArr, test.expectedArr) {
```
Consider using testify over reflect!!!

CameronSchafer replied:
Changed to testifies `Equalf()` method.

brvishnu92 commented on 02_bubble/CameronSchafer/bubble.go:
```diff
@@ -0,0 +1,95 @@
+		}
+		check = tempCheck
+	}
+	return s
```
Consider refactoring to return a copy of slice.

CameronSchafer replied:
Changed this function to sort a copy of the slice.

juliaogris commented on 02_bubble/CameronSchafer/bubble.go:
```diff
@@ -0,0 +1,98 @@
+	check := false
+
+	for !check {
+		tempCheck := true
```
just use `check` better name; `swapped` and invert logic.

CameronSchafer replied:
done 👍

juliaogris commented on 02_bubble/CameronSchafer/bubble.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// accidentally created an O(n2) complexity instead of O(n log n) complexity
+// merge sort: O(n2) sorting algorithm
```
haha :) - maybe you shouldn't call it merge sort then.
feel free to delete it, only bubble sort was required.

CameronSchafer replied:
haha I removed it 👍

juliaogris commented on 02_bubble/CameronSchafer/bubble_test.go:
```diff
@@ -0,0 +1,114 @@
+//bubble() test
+func TestBubbleOutput(t *testing.T) {
+	//test cases with descriptions.
+	testCases := []struct {
```
Consider using maps for your testCases with description as the key (see [Dave Cheney post](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests))

CameronSchafer replied:
done. 
Updated to use maps

juliaogris commented on 02_bubble/CameronSchafer/bubble.go:
```diff
@@ -0,0 +1,98 @@
+func merge(s []int) []int {
+	sBox := [][]int{}
+	for a := range s {
+		sBox = append(sBox, []int{s[a]})
```
use `make` if you know size of the slice to be generated (more runtime efficient).

CameronSchafer replied:
removed from code

CameronSchafer replied:
will keep in mind for future code 👍

juliaogris commented on 02_bubble/CameronSchafer/bubble.go:
```diff
@@ -0,0 +1,98 @@
+		var tBox [][]int
+		for i := 0; i < len(sBox); i += 2 {
+			// if the adjacent box exists then compare them
+			if i+1 != len(sBox) {
```
This is really hard to reason about.
I don't think you want to compare your loop index in every iteration. You could probably wrap the special case `i+1 == len(sBox)` somehow into the condition expression and pull it out of the loop - that would save you a nesting level.

CameronSchafer replied:
function removed from code.
will keep in mind for future code to reduce complexity

juliaogris reviewed (changes requested):
A few change requests around merge sort.
Please [limit the number of commits to one or two](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#limit-the-number-of-commits-to-one-or-two)

Commit:
Update test cases to use maps instead of structs
Update variable names to be more informative
Remove incorrect merge sort function

CameronSchafer committed:
Merge branch 'master' into lab2

juliaogris commented on 02_bubble/CameronSchafer/bubble.go:
```diff
@@ -0,0 +1,55 @@
+				}
+			}
+		}
+		check = swapped
```
check and swap do the same thing. Just use one.

juliaogris commented on 02_bubble/CameronSchafer/bubble.go:
```diff
@@ -0,0 +1,55 @@
+	for !check {
+		swapped := true
+		// sort once and then check
+		for i := 0; i < len(t); i++ {
```
Typically bubble sort's inner loop bound goes from 0 to i, with i being the out loop index ranging from n to 0.
After the first iteration of the outer loop the largest element in the slice has "bubbled up" to the last index and does not need to be considered for sorting any more (it already is sorted):
```go
	for i := len(s); i > 0; i-- {
		swapped := false
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
				swapped = true
			}
		}
		if !swapped {
			break
		}
	}
```
(feel free to use, adjust or ignore our reference solution).

juliaogris reviewed (approved):
LGTM,
though if you want to be perfectionist I think your bubble sort could still do with a little simplification and optimization.

CameronSchafer committed:
Merge branch 'master' into lab2
